### DISCUSSION TITLE
Non-modal dialogs and alert dialogs

### DISCUSSION THREAD

**Discussion by andrecasal** [adobe/react-spectrum#5981] [Q&A]
Dialogs can be non-modal, like a Chat box (non-modal dialog, can be dismissed/minimized) or a Cookies banner (non-modal alert dialog, can't be dismissed until an answer is given).

The docs say a Dialog needs to go inside a Modal or a Popover, so I'm wondering if it's possible to create non-modal dialogs and how. I've tried removing the Modal wrapper from a Dialog, but that just renders the dialog below the trigger.

**Comment by snowystinger** [comment]
For both those examples, I think they'd fall in this category https://github.com/adobe/react-spectrum/discussions/6000#discussioncomment-8671656

You could checkout how we implemented Toasts in React Spectrum https://react-spectrum.adobe.com/react-spectrum/Toast.html

**Comment by andrecasal** [comment]
The `data-react-aria-top-layer` attribute seems to elevate an element to be on top of everything else, including the modal. For something like a chat box or a cookies banner this would have to be applied to the body, which is probably not the right solution.

I haven't found the`data-react-aria-top-layer` attribute anywhere on the Toast, but I'll try to figure out how Spectrum implements non-modal dialogs like Toasts.

Thanks @snowystinger!

**Reply by devongovett** [reply]
No it should not go on the body, it should go on the chat modal itself. If you can't add something to that element since it's rendered by a third party, you might be able to control which element that window gets rendered into. If so, you could create a new element inside the body with data-react-aria-top-layer and render the chat window inside that.

**Reply by andrecasal** [reply]
It seems I may not have explained myself correctly @devongovett.

In a situation where a user needs live support to figure something out on the interface, clicks a fixed positioned button to open a live chat, and is still able to interface with the rest of the app. How would one handle such a dialog that doesn't trap focus and has no semi-transparent underlay?

**Reply by devongovett** [reply]
Right, you'd add data-react-aria-top-layer to the button + chat window. Does that not work for you?

**Reply by andrecasal** [reply]
Thanks @devongovett, I got it to work partially:
1. The trigger button has data-pressed="true" while the dialog is opened, giving it the wrong styles
2. Scroll interaction is still disabled for the rest of the page

**Reply by andrecasal** [reply]
@devongovett because `<Modal />` doesn't span the entire viewport, we can click on elements outside the dialog, but focus is still trapped, so we can't tab outside the Dialog and scrolling is disabled.

Adding `data-react-aria-top-layer` doesn't seem to help.

```tsx
<p className="hover:text-emerald-300 cursor-pointer">Test element</p>
<DialogTrigger>
	<Button className="rounded-lg bg-brand p-3 text-background hover:bg-red-500" data-react-aria-top-layer>Open Dialog</Button>
	<Modal>
		<Dialog className="fixed left-1/2 top-1/2 rounded-2xl border bg-background p-6 shadow-sm" data-react-aria-top-layer>
			{({ close }) => (
				<>
					<h1>Dialog</h1>
					<p>Awesome paragraph</p>
					<Button className="rounded-lg bg-brand p-3 text-background" onPress={close} autoFocus>
						Close Dialog
					</Button>
				</>
			)}
		</Dialog>
	</Modal>
</DialogTrigger>
```

**Reply by snowystinger** [reply]
I think there was a little miscommunication here. I don't think your chat would be a dialog, or even a modal. It's just part of your app's normal usage and it happens to display over some other things. If you want to be able to interact with it while one of our Dialogs is open as well, then you'd need the `data-react-aria-top-layer`, as you have in the code you shared.

You could potentially use our positioning code and detection for interact outside. These aren't documented as we do sometimes make changes to them. It might be worth a discussion for something like this.

I'd also structure it such that the chat overlay is a `landmark` for ease of navigation.

@devongovett may have some other suggestions for this as well.

**Reply by andrecasal** [reply]
This could be done by hand, but we'd have to code all the accessibility, focus management, etc ourselves.

Wouldn't it make sense for a chatbox to be a non-modal dialog?

**Reply by snowystinger** [reply]
what is a non-modal dialog? isn't it just a piece of your normal UI that you're displaying in a different spot than the document flow?
the only focus management for it should be to make it a landmark. otherwise, it doesn't trap focus, and you can get restore focus by adding a `<FocusScope restoreFocus>`
